### PR TITLE
Improve test code and structure

### PR DESCRIPTION
**Pull request by fobilow**
- moved all test into _test package
    - moved all helper funcs into all_test.go
    - neater teardown pattern
    - added flags to make tests more interactive
    - refactored some methods to make them testable
    - stop storing mails on disk, gitdb now keep them in memory instead

### CODE REVIEW

File: all_test.go [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+	flag.BoolVar(&flagFakeRemote, "fakerepo", true, "create fake remote repo for tests")
+	flag.Parse()
+
+	gitdb.SetLogLevel(gitdb.LogLevel(flagLogLevel))
+	m.Run()
+}
+
+func setup(t testing.TB) func(t testing.TB) {
+	//fail test if git is not installed
+	if _, err := exec.LookPath("git"); err != nil {
```
Comment (cubixle) [inline all_test.go]: maybe make this part of TestMain so you are not relying on it for every test.

File: all_test.go [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+		return
+	}
+}
+
+func getDbConn(t testing.TB) *gitdb.Connection {
+	conn, err := gitdb.Open(getConfig())
+	if err != nil {
+		t.Errorf("getDbConn failed: %s", err)
+	}
+	fmt.Println("test db connection opened")
```
Comment (cubixle) [inline all_test.go:89]: maybe a  debug log?

File: all_test.go [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+		println(v)
+		println(err.Error())
+		want = 0
+	}
+
+	return want
+}
+
+func generateInserts(t testing.TB, count int) {
+	fmt.Printf("inserting %d records\n", count)
```
Comment (cubixle) [inline all_test.go:195]: debug log?

File: all_test.go [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+	for i := 0; i < count; i++ {
+		if err := testDb.Insert(getTestMessage()); err != nil {
+			t.Errorf("generateInserts failed: %s", err)
+		}
+
+	}
+	fmt.Println("done inserting")
```
Comment (cubixle) [inline all_test.go:202]: debug log?

File: db_test.go [hunk 5]
```diff
@@ -1,226 +1,17 @@
...
 
 func TestMigrate(t *testing.T) {
-	setup()
-	defer testDb.Close()
+	teardown := setup(t)
+	defer teardown(t)
 
 	m := getTestMessageWithId(0)
-	insert(1)
+	insert(m, false)
```
Comment (cubixle) [inline db_test.go]: returned error not checked

File: git.go [hunk 6]
```diff
@@ -67,9 +67,10 @@ func (g *gitdb) gitClone() error {
 			notification := "Contact your database admin to add your public key to git server\n"
 			notification += "Public key: " + fmt.Sprintf("%s", fb)
 
+			log(notification)
 			logTest(notification)
 
-			err = newMail(g, "Database Setup Error", notification).send()
+			g.sendMail(newMail("Database Setup Error", notification))
```
Comment (cubixle) [inline git.go:66]: err not checked. missing err var assignment.

File: schema_test.go [hunk 7]
```diff
@@ -1,19 +1,20 @@
...
 func TestNewAutoBlock(t *testing.T) {
-	setup()
-	defer testDb.Close()
-
-	insert(1)
+	teardown := setup(t)
+	defer teardown(t)
 
 	m := getTestMessage()
+	insert(m, false)
```
Comment (cubixle) [inline schema_test.go]: err not checked

File: ssh_test.go [hunk 8]
```diff
@@ -1,9 +1,9 @@
-package gitdb
+package gitdb_test
 
 import (
 	"testing"
 )
 
-func TestGenerateSSHKeyPair(t *testing.T){
+func TestGenerateSSHKeyPair(t *testing.T) {
```
Comment (cubixle) [inline ssh_test.go]: maybe remove this unless a test is planned to go here.

File: write_test.go [hunk 9]
```diff
@@ -84,45 +33,45 @@ func TestInsertMany(t *testing.T) {
...
 func TestDelete(t *testing.T) {
-	setup()
-	defer testDb.Close()
-
-	insert(1)
+	teardown := setup(t)
+	defer teardown(t)
 
 	m := getTestMessageWithId(0)
+	insert(m, flagFakeRemote)
```
Comment (cubixle) [inline write_test.go]: err not checked

File: ssh_test.go (hunk 8 above)
Comment (fobilow) [reply ssh_test.go]: yea there is a plan to write the test

File: all_test.go (hunk 1 above)
Comment (fobilow) [reply all_test.go]: good shout! thanks

File: git.go (hunk 6 above)
Comment (fobilow) [reply git.go:66]: nice spot! sendMail appends to array and always returns nil. Will fix the api for the method instead. Thanks

File: ssh_test.go (hunk 8 above)
Comment (fobilow) [reply ssh_test.go]: on second thought I think you are right, the ssh funcs are all unexported and can't be unit tested so will drop the test file. Cheers

File: all_test.go (hunk 2 above)
Comment (fobilow) [reply all_test.go:89]: these are quite useful for now so will keep them and revisit in the future thanks

**Pull request merged**